pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+
+                    // Set temporary value for Pj as global index, j. Will be mapped to
+                    // appropriate coarse-grid column index after all data is filled in. 
+                    Pj[nnz] = Sj[jj];
+                    I j = Sj[jj];
+
+                    // Initialize numerator as a_ij
+                    T numerator = Sx[jj];
+
+                    // Sum over strongly connected fine points
+                    for (I kk = Sp[i]; kk < Sp[i+1]; kk++) {
+                        if ( (splitting[Sj[kk]] == F_NODE) && (Sj[kk] != i) ) {
+                            
+                            // Get column k and value a_ik
+                            I k = Sj[kk];
+                            T a_ik = Sx[kk];
+
+                            // Get a_kj (have to search over k'th row in A for connection a_kj)
+                            T a_kj = 0;
+                            T a_kk = 0;
+                            for (I search_ind = Ap[k]; search_ind < Ap[k+1]; search_ind++) {
+                                if (Aj[search_ind] == j) {
+                                    a_kj = Ax[search_ind];
+                                }
+                                else if (Aj[search_ind] == k) {
+                                    a_kk = Ax[search_ind];
+                                }
+                            }
+
+                            // If sign of a_kj matches sign of a_kk, ignore a_kj in sum
+                            // (i.e. leave as a_kj = 0)
+                            if (signof(a_kj) == signof(a_kk)) {
+                                a_kj = 0;
+                            }
+
+                            // If a_kj == 0, then we don't need to do any more work, otherwise
+                            // proceed to account for node k's contribution
+                            if (std::abs(a_kj) > 1e-16) {
+                                
+                                // Calculate sum for inner denominator (loop over strongly connected C-points)
+                                T inner_denominator = 0;
+                                for (I ll = Sp[i]; ll < Sp[i+1]; ll++) {
+                                    if (splitting[Sj[ll]] == C_NODE) {
+                                        
+                                        // Get column l
+                                        I l = Sj[ll];
+                                        
+                                        // Add connection a_kl if present in matrix (search over kth row in A for connection)
+                                        // Only add if sign of a_kl does not equal sign of a_kk
+                                        for (I search_ind = Ap[k]; search_ind < Ap[k+1]; search_ind++) {
+                                            if (Aj[search_ind] == l) {
+                                                T a_kl = Ax[search_ind];
+                                                if (signof(a_kl) != signof(a_kk)) {
```

Comment:
This could be changed to something like
`if( modified && (signof(a_kl) == signof(a_kk)) {`
`  a_kl = 0;`
`}`
`inner_denominator += a_kl;`